### PR TITLE
VSI-ETH buy and sell

### PR DESCRIPTION
**Pull request by mikec**
implements functions to allow 1:1 buy/sell exchange of VSI-ETH and ether.

ether is held in VersiPool, which has VersiEtherToken set as an owner. VersiEtherToken proxies ether from the buyer/seller to and from VersiPool.

### CODE REVIEW

File: contracts/VersiEtherToken.sol [hunk 1]
```diff
@@ -1,30 +1,60 @@
...
+    require(msg.sender == address(pool));
+    require(msg.value > 0);
+  }
+
+  /// @dev Buys VSI-ETH
+  function buy() public payable {
+    require(msg.value > 0);
+    balances[msg.sender] = balances[msg.sender].add(msg.value);
+    totalTokens = totalTokens.add(msg.value);
+    pool.deposit.value(msg.value)();
```
Comment (jpkcambridge) [inline contracts/VersiEtherToken.sol]: Can you help me understand this syntax more.  Why are there two sets of parentheses, and why is the format different than the withdraw action on line 42?

**Comment by cwhinfrey** [comment]
@mikec Would it make more sense for Versi to own VersiPool instead of VersiEtherToken? I was thinking buy and sell could be implemented on the Versi contract which owns both the pool and VersiEtherToken. The main Versi contract could then use the create and destroy functions on VersiEtherToken in it's implementation of buy and sell.

File: contracts/VersiEtherToken.sol (hunk 1 above)
Comment (cwhinfrey) [reply contracts/VersiEtherToken.sol]: @jpkcambridge the `.value(VALUE)` is added to set the amount of ETH to be sent with the call. So to call deposit without any ETH would just be `pool.deposit();` but to send some ETH we add `.value(msg.value)` after deposit. We don't send any ETH to the withdraw function so it's just `pool.withdraw(value)` where `value` is a parameter of `withdraw`

**Comment by mikec** [comment]
@cwhinfrey my thinking here was to separate public buy/sell from mint/destroy. The mint/destroy process doesn't have to touch VersiPool, since there are no funds being transferred. The flow for public buy/sell would be `randomAddress <-> VersiEtherToken <-> VersiPool` and the flow for VSI-ETH mint/destroy would be `Versi -> VersiEtherToken`.

**Comment by cwhinfrey** [comment]
@mikec Agreed that the mint/destroy process doesn't have to touch `VersiPool`. But the buy/sell functions mint and destroy tokens and should probably call those functions to avoid repeating logic. What's the downside to having Versi own VersiPool and use `issueVersiEtherToken` and `redeemVersiEtherToken` like we had stubbed out?

**Comment by cwhinfrey** [comment]
The way we had it stubbed keeps us from having to have multiple owners on VersiPool, VersiEtherToken doesn't need to know about VersiPool, and keeps the logic out of the ERC20 contract which I think makes sense if ERC20's are hard to upgrade.

**Comment by cwhinfrey** [comment]
I made a PR against this branch to show what I'm trying to say https://github.com/levelkdev/versi-contracts/pull/4/files

**Comment by mikec** [comment]
@cwhinfrey I read through the PR and understand the approach. I'm worried about the direct owner relationship between Versi and VersiPool, because Versi might have a lot more logic as we build this out. It feels dangerous to let a contract with functions not related to the pool have potential access to the pool. Opens up more possibilities for us to make a mistake.

Also, think about how we will have to add new `VSI-XXX` tokens - we have to add new issue/destroy functions to Versi.sol  for every `VSI-`. I agree there's some repeated logic in VersiEtherToken right now, but we can abstract this into internal functions within VersiEtherToken.

In short, I feel like it's safer to have multiple tightly scoped owners on the pool, rather than one loosely scoped owner.

**Comment by mikec** [comment]
I moved the logic to internal function in the last commit.

**Comment by mikec** [comment]
Just thought of something else. If we end up having multiple `VSI-XXX` tokens, we probably want them to have separate "pools". For example, if we have `VSI-ETH` and `VSI-OMG`, we might have one pool that holds ether and another that holds OMG (via a balance on OMG's ERC20). We should talk through the implications of this. I'm assuming it's possible that we launch with VSI-ETH only, so we should have a clear migration path for adding additional tokens, if that's something we think we'll need.

**Comment by mikec** [comment]
closing in favor of https://github.com/levelkdev/versi-contracts/pull/5
